cdap-api/src/main/java/co/cask/cdap/internal/api/DefaultDatasetConfigurer.java

```diff
@@ -50,21 +52,43 @@
   }
 
   public void addStreams(Map<String, StreamSpecification> streams) {
+    Set<String> duplicateStreams = findDuplicates(streams.keySet(), this.streams.keySet());
+    checkArgument(duplicateStreams.isEmpty(),
+                  "Streams %s have been added already. Remove the duplicates.", duplicateStreams);
     this.streams.putAll(streams);
   }
 
   public void addDatasetSpecs(Map<String, DatasetCreationSpec> datasetSpecs) {
+    Set<String> duplicateDatasetSpecs = findDuplicates(datasetSpecs.keySet(), this.datasetSpecs.keySet());
+    checkArgument(duplicateDatasetSpecs.isEmpty(),
+                  "Dataset Instances %s have already been added. Remove the duplicates.", duplicateDatasetSpecs);
     this.datasetSpecs.putAll(datasetSpecs);
   }
 
   public void addDatasetModules(Map<String, String> datasetModules) {
+    for (Map.Entry<String, String> newEntry : datasetModules.entrySet()) {
+      if (datasetModules.containsKey(newEntry.getKey())) {
```

Comment:
Yes, good catch! Thanks! Updated the PR and also enhanced the unit test to cover this case.